### ISSUE TITLE
index out of range when using splat square bracket indexing in template_file input

### ISSUE THREAD

**Issue by sjwl** [hashicorp/terraform#13828]
### Terraform Version
0.9.3

### Affected Resource(s)
All

### Terraform Configuration Files
```hcl
variable count { default = 1 }

resource "test_resource" "foo" {
  count = "${var.count}"
  required = "foo"

  required_map {
    bar = "bar"
  }

}

data "template_file" "my_script" {
  template = "$${x}:1234"
  count    = "${var.count}"

  vars {
  x = "${test_resource.foo.*.required[count.index]}"
  }
}
```

### Expected Behavior
Changing count up or down should add or remove resources accordingly.

### Actual Behavior
```
Error refreshing state: 1 error(s) occurred:

* data.template_file.my_script: 1 error(s) occurred:

* data.template_file.my_script[1]: index 1 out of range for list test_resource.foo.*.required (max 1) in:

${test_resource.foo.*.required[count.index]}
```
### Steps to Reproduce

1. `TF_VAR_count=1 terraform apply`
2. `TF_VAR_count=2 terraform plan`

The first step works, but the error happens on the second step.  Same error if the count is increasing or decreasing.

### Important Factoids
~~The workaround is to use `element()` instead of square bracket notation~~

**Comment by vancluever** [comment]
Could possibly be related to #13793.

One thing to note about `element()` is that the function by design wraps the list, so while you might not get any errors, you might not be getting the element that you want - @sjwl can you check to see if you are getting the intended output from the your plan in light of this?

**Comment by sjwl** [comment]
@vancluever you're right.  unfortunately `element()` workaround doesn't actually interpolate the expected values.  :-(

**Comment by vancluever** [comment]
@sjwl if you have the capacity to do so, you might want to try building a local copy of TF with the change detailed in #13793 included to see if it fixes things for you - looks like we are getting success with this in other cases (ie: #13867).

**Comment by sjwl** [comment]
@vancluever I tried it, and looked promising at first.  

1. `TF_VAR_count=1 terraform apply`  (WORKED, as expected)
2. `TF_VAR_count=2 terraform apply` (WORKED, yah)
3. `TF_VAR_count=3 terraform apply` (ERROR)
```
Error refreshing state: 1 error(s) occurred:

* data.template_file.my_script: 1 error(s) occurred:

* data.template_file.my_script[2]: index 2 out of range for list test_resource.foo.*.required (max 2) in:

${test_resource.foo.*.required[count.index]}
```

**Comment by vancluever** [comment]
@sjwl confirmed - looking into it now - thanks!

**Comment by vancluever** [comment]
Alright I think I found the issue with this one - the issue being that during a refresh, new *resources* do not exist in the graph yet, while data sources, by design, do: meaning that the new data source added by the expansion with the updated count by `ResourceCountTransformer` in `NodeRefreshableDataResource` does not have a corresponding `test_resource` instance in the graph to index against during refresh.

This relationship is actually lot more clearer if you base the count of the data source on the count of the resource, at which point you get `value of 'count' cannot be computed` when trying to do a run.

To fix this, preliminarily, I added a `ConfigTransformer` to the refresh builder that adds new resources from the config as `NodePlannableResource`s with a filter of `config.ManagedResourceMode`:

```
// Creates all the resources that aren't in the state
&ConfigTransformer{
  Concrete:   concretePlannableResource,
  Module:     b.Module,
  Unique:     true,
  ModeFilter: true,
  Mode:       config.ManagedResourceMode,
},
```

(The concretePlannableResource is basically just a function that returns a `NodePlannableResource`)

Applies pass after this.

Doing this did not seem to break any existing tests in the `terraform` package, but I still have yet to review the implications of adding `NodePlannableResource`s to the graph and seeing if we ultimately get the same end results. One question I would be asking is if `test_resource.foo.*.required` is computed, will the data source fire at the right time in the graph to ensure that the correct data in the data source was available?

**Comment by vancluever** [comment]
Quick update on this one before I leave it for now: changing the `test_resource` to `random_id` and then trying to base some outputs on the data did seem to work, but when trying to resource -> data source I got another `count cannot be computed` error. Switching the count back to the variable worked though!

**Comment by vancluever** [comment]
Another update: out of curiosity I decided to try this again with basing the child data source's count off of a *non-computed* attribute in the resource instance. And it worked!

So this setup seems to work just fine, and as long as you are not basing a computed `count` of an attribute that is not going to be known by the time you hit plan (or refresh on the second apply), you should be fine.

Working config for reference:

```
variable count {
  default = 1
}

resource "random_id" "foo" {
  count       = "${var.count}"
  byte_length = 4
}

data "template_file" "my_script" {
  template = "$${x}:1234"
  count    = "${length(random_id.foo.*.byte_length)}"

  vars {
    x = "${random_id.foo.*.hex[count.index]}"
  }
}

output "out" {
  value = "${join(",", data.template_file.my_script.*.rendered)}"
}
```

Will work on trying to formalize this into #13793 or a different PR soon.

**Comment by vancluever** [comment]
Alright, another update before I finish for the morning - while this works for *adding* resources, this does not necessarily work for *removing* resources as a result of a negative count change (going from 4 resources to 2, for example).

The issue seems to be that while the resources exist in state, they do not exist in the graph by the time the data sources are evaluated and we run into the same `index out of range` issue.

At this point I'm trying to think what the best way would be to support this behaviour. This particular use case does seem to work as expected for creating resources, so I think it would be fine to support it going the other way around. I almost wonder if the easiest way to support this behaviour would be to create a new dynamic expansion refresh node for resources that only did one-way expansion of count on both resources and data sources, leaving more detailed reconciliation to the plan graph. Since refresh is only supposed to be updating resources known in the state, this makes sense to me.

**Comment by vancluever** [comment]
Okay, final-ish assessment:

Scaling in resources in the fashions mentioned above is problematic not because of the state actually, but more because of the fact that interpolations for values rely on raw configuration data, and not necessarily state. The bounds checking is using the config as it is interpolated in any current run as the source of truth, and not necessarily the state.

As such, I think that the best way to fix this is to not use the state as the base for the refresh graph. This seems to introduce orphans too early on in the process for them to be properly identified as such and they wouldn't get transformed properly, or at least not with the state the refresh builder is in right now. What I would imagine would need to ultimately happen is that the graph will look fairly similar to a plan graph with a few provisions for anything that specifically needs to happen during the refresh phase versus the plan phase. Config needs to be the source of truth so that reasonably intuitive interpolation decisions can be made, and both resources and data source need expandable nodes so that a graph state that represents the current state of interpolation can be created. I would posit that orphans should not even be walked at all, as they are likely to contain stale raw config data that could not be reasonably interpolated.

**Comment by vancluever** [comment]
Alright, I think I got a good fix pinned down. @sjwl feel free to give the referenced PR for a spin, still have to work on tests but I have been using your test case as a repro. Let me know if you get a chance to try it out!

**Comment by apparentlymart** [comment]
Thanks to great work from @vancluever we now have a fix for this in master, which should be included in the next release. Thanks, @vancluever!

**Comment by vancluever** [comment]
@apparentlymart happy to help as always and glad to see it's fixing things! 😁

**Comment by Quentin-M** [comment]
Hell yeah! Thanks!

**Comment by Uncurlhalo** [comment]
I see the change posted earlier has been merged into master. I'm on terraform version 0.9.11 and seem to still be getting this issue. Will it not be merged in until version 0.10?

**Comment by aabdullah-bos** [comment]
It looks like the change submitted by @vancluever isn't available in terraform `0.11.x`. As I am running terraform `0.11.11` and I am still seeing the behavior described in the initial comment/description. I tried a test case in `0.12.0-beta2` and it looks like the issue is resolved, however, I need to use the PagerDuty integration and it looks like the development snapshot isn't working correctly.

Can someone point me to a workaround until `0.12.0` is released?
